### PR TITLE
feat(stage_def): add type-safe Dep/Out descriptors

### PR DESCRIPTION
**Pull request by sjawhar**
## Summary

- Replace nested `class deps:`/`class outs:` syntax with `Dep[T]`/`Out[T]` descriptors
- Provides proper type inference when accessing loaded dependencies and outputs
- Override `model_dump()` to exclude descriptor fields from JSON serialization
- Add `output_queue` fixture to reduce executor test boilerplate
- Delete exploratory `type_safety_final.py` file

## Example

```python
class ProcessParams(StageDef):
    data: Dep[pandas.DataFrame] = Dep("input.csv", loaders.CSV())
    result: Out[dict[str, int]] = Out("output.json", loaders.JSON())
    threshold: float = 0.5

@stage()
def process(params: ProcessParams) -> None:
    df = params.data  # Type: pandas.DataFrame (inferred!)
    params.result = {"count": len(df)}  # Type-checked assignment
```

## Test plan

- [x] All 2258 tests pass
- [x] Type checking clean with basedpyright
- [x] Added pickle tests for StageDef classes
- [x] Added CLI integration test for end-to-end StageDef usage

🤖 Generated with [Claude Code](https://claude.com/claude-code)

### CODE REVIEW

**Review by sjawhar — COMMENTED** [review]
## Review Summary

**Recommendation: Request Changes** - There is one blocking issue that must be fixed before merge.

## What Works Well

- **Clean API simplification**: The new `Dep[T]`/`Out[T]` descriptor syntax is more intuitive than the nested class approach. Accessing `params.data` directly instead of `params.deps.data` is cleaner.

- **Proper type safety**: The generic descriptors provide type inference at access time, which was the goal of this PR.

- **Good inheritance handling**: The `__set_name__` approach with spec copying in both the descriptor and `__init_subclass__` ensures inheritance works correctly.

- **Comprehensive test coverage**: New pickle tests verify multiprocessing compatibility, inheritance tests cover edge cases, and the CLI integration test validates end-to-end behavior.

- **Code simplification**: Removing WeakKeyDictionary, namespace classes, and complex descriptor machinery in favor of storing data directly in `_deps_data`/`_outs_data` dicts is a good simplification.

- **README updated**: Documentation reflects the new API correctly.

## Blocking Issues

### 1. `model_dump()` fails when user passes `exclude` parameter

**Location**: `src/pivot/stage_def.py` line 200-203

The current implementation:
```python
def model_dump(self, **kwargs: Any) -> dict[str, Any]:
    descriptor_names = set(self._deps_specs.keys()) | set(self._outs_specs.keys())
    return super().model_dump(exclude=descriptor_names, **kwargs)
```

This will raise `TypeError: got multiple values for keyword argument 'exclude'` if a user calls `params.model_dump(exclude={'threshold'})`.

**Fix**: Merge user-provided exclude with descriptor names:
```python
def model_dump(self, **kwargs: Any) -> dict[str, Any]:
    descriptor_names = set(self._deps_specs.keys()) | set(self._outs_specs.keys())
    user_exclude = kwargs.pop("exclude", None)
    if user_exclude:
        descriptor_names |= set(user_exclude) if isinstance(user_exclude, (set, frozenset)) else {user_exclude}
    return super().model_dump(exclude=descriptor_names, **kwargs)
```

**Add test**: Should test that `model_dump(exclude={'some_param'})` works correctly.

## Important Issues

None - the implementation is otherwise solid.

## Additional Suggestions

### 1. Consider `model_dump_json()` override

Pydantic also has `model_dump_json()` method. If users call this directly (rather than going through `model_dump()`), descriptor fields might be included. Consider overriding it as well for consistency, or verify Pydantic delegates to `model_dump()` internally.

### 2. Minor: pyright ignore comments could be more specific

The comments like `# pyright: ignore[reportAttributeAccessIssue,reportUnknownMemberType,reportUnknownVariableType]` are quite long. While they work, documenting WHY each ignore is needed (as per CLAUDE.md) would help future maintainers understand.

## Testing Notes

- All 2257 tests pass (excluding 4 pre-existing fingerprint test failures on main)
- Type checking passes with basedpyright
- Linting passes with ruff
- New tests cover:
  - Pickle roundtrip for StageDef with deps/outs
  - Inheritance (child inherits parent's deps/outs)
  - CLI integration with StageDef
  - output_queue fixture for test boilerplate reduction

## Next Steps

1. **Fix the `model_dump()` exclude parameter handling** (blocking)
2. Add a test for `model_dump(exclude=...)` behavior
3. Optionally consider `model_dump_json()` override

File: src/pivot/stage_def.py [hunk 1]
```diff
@@ -1,348 +1,264 @@
...
-    # Add __init__ that takes parent reference
-    def namespace_init(self: _NamespaceInstance) -> None:
-        self._parent_stage_def = None  # pyright: ignore[reportPrivateUsage] - protocol attribute
+    def __init__(self, path: str, loader: loaders.Loader[T]) -> None:
+        self.path = path
+        self.loader = loader
+        self._name = ""
 
-    namespace_dict["__init__"] = namespace_init
+    def __set_name__(self, owner: type[StageDef], name: str) -> None:
```
Comment (sjawhar) [inline src/pivot/stage_def.py]: :+1: Good use of `__set_name__` for Out descriptors as well, maintaining consistency with the Dep pattern.

File: src/pivot/stage_def.py [hunk 2]
```diff
@@ -1,348 +1,264 @@
...
 
-    # Use parameterized descriptors for deps/outs access
-    deps: ClassVar[_NamespaceDescriptor] = _NamespaceDescriptor(
-        "_deps_namespace_cls", "_deps_instance", "dependencies"
-    )
-    outs: ClassVar[_NamespaceDescriptor] = _NamespaceDescriptor(
-        "_outs_namespace_cls", "_outs_instance", "outputs"
-    )
+    _deps_data: dict[str, Any] = pydantic.PrivateAttr(default_factory=dict)
+    _outs_data: dict[str, Any] = pydantic.PrivateAttr(default_factory=dict)
```
Comment (sjawhar) [inline src/pivot/stage_def.py]: :+1: Proper `__init_subclass__` usage to ensure inheritance works correctly - child classes get copies of parent specs rather than sharing the same dict.

File: tests/test_stage_def.py [hunk 3]
```diff
@@ -391,3 +390,75 @@ class ProcessParams(stage_def.StageDef):
...
     with pytest.raises(pydantic.ValidationError):
         ProcessParams(threshold=2.0)
+
+
+# ==============================================================================
+# Inheritance tests
+# ==============================================================================
+
+
+def test_stage_def_child_inherits_parent_deps(tmp_path: pathlib.Path) -> None:
```
Comment (sjawhar) [inline tests/test_stage_def.py]: :+1: Good coverage of inheritance behavior. Testing that child classes properly inherit parent deps/outs is important for the descriptor pattern.
